Split build workflow into composite actions

getchoo opened the pull request:
~~blocked on #3702~~

The main point of this is to:

- Make our main CI workflow *much* more maintainable by splitting the beast into nice, easily digestible chunks 
- Encourage re-use and compartmentalization of our workflow steps,
- Enable easy global changes to variables like our Qt version and modules
- Make exceptions for different platforms much more obvious (again, like Qt version)

**ccache also works with CodeQL and MSVC now!**

getchoo committed:
ci: split build workflow into composite actions

Signed-off-by: Seth Flynn <getchoo@tuta.io>

getchoo committed:
ci: run build workflow directly on push/prs

Calling this from another workflow only for these events doesn't make
much sense now

Signed-off-by: Seth Flynn <getchoo@tuta.io>

getchoo committed:
ci: trigger_release -> release

Signed-off-by: Seth Flynn <getchoo@tuta.io>

Ryex reviewed (commented):
I'm very much liking this

getchoo committed:
ci(codeql): use setup-dependencies action

Signed-off-by: Seth Flynn <getchoo@tuta.io>

getchoo committed:
ci: use sccache on windows

This apparently works with less hacks, and is actually suggested by the
action we're using

Signed-off-by: Seth Flynn <getchoo@tuta.io>

getchoo commented:
Checked all Windows binaries (aside from ARM, I could only check that it was indeed ARM). Release builds will need to be tested after this is landed via workflow_dispatch, but it'll probably work
